Refactor checkMaxConnections

devinbileck opened the pull request:
Fix connection limit checks so as to prevent the following warning:

> WARN  b.n.p2p.peers.PeerManager: No candidates found to remove (That
case should not be possible as we use in the last case all
connections).

mrosseel commented:
this might be a good opportunity to add a unit test for PeerManager (possibly limited to the scope of connection handling)

ripcurlx reviewed (dismissed):
utACK

devinbileck commented:
> this might be a good opportunity to add a unit test for PeerManager (possibly limited to the scope of connection handling)

It seems the PeerManagerTest is ignored and in need of an overhaul.

ripcurlx commented:
Could someone of the exisiting contributors try out this branch before merging, as it would cause big problems if some use-case handling is wrong. I just read through the code and it looks fine, but it would be good to give this a local test as well.

ManfredKarrer commented:
I think that code part is really a good candidate for a test. The logic is a bit involved and bigger changes here are tough to test manually. I would postpone otherwise merge until someone had dedicated enough time to be sure all is correct.

devinbileck commented:
Agreed. I will try to look into implementing some tests. Just not familiar with this area so will need some time to ramp up.

ripcurlx commented:
> Agreed. I will try to look into implementing some tests. Just not familiar with this area so will need some time to ramp up.

@devinbileck Are you still on this PR?

devinbileck commented:
I started looking into it but didn't make much progress. I haven't focused on it very much lately. If someone else is able to contribute something, feel free. Otherwise I will try to get to it.

devinbileck committed:
Change access level for checkMaxConnections to be tested

devinbileck committed:
Add MockNode that allows for simulating connections

devinbileck committed:
Add PeerManagerTest

The old PeerManagerTest was located under network/p2p/routing, which is
no longer the correct location. Additionally, it was outdated so I
just removed it and added a new file under network/p2p/peers containing
tests for checkMaxConnections.

devinbileck committed:
Refactor checkMaxConnections

Fix connection limit checks so as to prevent the following warning:

> WARN  b.n.p2p.peers.PeerManager: No candidates found to remove (That
case should not be possible as we use in the last case all
connections).

ripcurlx commented:
@devinbileck Could you please resolve the conflict and add some tests?
@freimair Could you review the code?

devinbileck commented:
I have added some tests. I will be committing them shortly.

devinbileck committed:
Add testCompile dependency to core

This is necessary because bisq.network.p2p.MockNode imports
bisq.core.network.p2p.seed.DefaultSeedNodeRepository.

devinbileck commented:
@ripcurlx @freimair ready for review.

ripcurlx commented:
@freimair As I haven't touched this part of the codebase so far, could you please do the review? Thanks!

freimair commented on build.gradle:
```diff
@@ -205,6 +205,7 @@ configure(project(':common')) {
 configure(project(':p2p')) {
     dependencies {
         compile project(':common')
+        testCompile project(':core')
```
```suggestion

```

freimair commented on p2p/src/test/java/bisq/network/p2p/MockNode.java:
```diff
@@ -0,0 +1,86 @@
+
+package bisq.network.p2p;
+
+import bisq.core.network.p2p.seed.DefaultSeedNodeRepository;
```
```suggestion
import bisq.network.p2p.seed.SeedNodeRepository;
```

freimair commented on p2p/src/test/java/bisq/network/p2p/MockNode.java:
```diff
@@ -0,0 +1,86 @@
+        this.maxConnections = maxConnections;
+        networkNode = mock(NetworkNode.class);
+        Storage<PeerList> storage = new Storage<>(mock(File.class), mock(PersistenceProtoResolver.class), mock(CorruptedDatabaseFilesHandler.class));
+        peerManager = new PeerManager(networkNode, mock(DefaultSeedNodeRepository.class), new ClockWatcher(), maxConnections, storage);
```
```suggestion
        peerManager = new PeerManager(networkNode, mock(SeedNodeRepository.class), new ClockWatcher(), maxConnections, storage);
```

freimair reviewed (changes requested):
IMO we could get away with mocking the `SeedNodeRepository` superclass, thus eliminating the dependency to `core`. Can you verify that this is correct?

devinbileck committed:
Update based on review feedback

Mock the SeedNodeRepository superclass, thus eliminating the dependency
to core.

devinbileck commented:
Yes, good point. Updated!
